src/grandorgue/dialogs/GOOrganSettingsDialog.cpp

```diff
@@ -682,6 +708,7 @@ void GOOrganSettingsDialog::Load(bool isForce) {
     float autoTuningCorrection = m_Last->config->GetAutoTuningCorrection();
     unsigned delay = m_Last->config->GetDelay();
     unsigned releaseLength = m_Last->node->GetEffectiveReleaseTail();
+    int toneBalance = m_Last->config->GetToneBalanceValue();
```

Comment:
Both variants bring the same GO behavior. The differnce is only visual if the value is not set at the pipe level: in your approach the user can not see the value that is actually used, but in my approach he/she does.

The same overriding logic exists with the `releaseTail` (one line above). Now the uses can see the actual working value. I suggest to make the same with the `toneBalance`. But if you do not want to do it, I won't insist.

>  the behaviour would be more intuitive

It is more intuitive for a programmer who understands how the program works. But it is not intuitive for the user, that would prefer some WYSIWYG approach.

Reply:
> But it is not intuitive for a musician, that would prefer some WYSIWYG approach.

I cannot say for sure, I'm primarily a musician that has learned some programming! However, the same method I propose is used for tuning and gain/amplitude level adjustments too. For now I'd stay with that approach.

Reply:
> the same method I propose is used for tuning and gain/amplitude level adjustments too

But tuning and gain/amplitude level have an additive behavior in opposite to the overriding behavior of tune balance. Displaying zero is ok for an additive thing, but not for an overriding thing.

Reply:
> Displaying zero is ok for an additive thing, but not for an overriding thing.

Perhaps that's a sound reasoning. But then we _must_ re-design the behaviour of the value and effective value calculations. If we enforce a writing of the value to children like you suggest - then setting 0 for a child should _not_ fall back again to the parent value but instead really set the 0 value for that child and thus always override any parent value.

Reply:
As I see no way to determine if a child intentionally sets its value to 0 wishing to override parent value, basically the whole get effective value reduces to always just getting the current item value. When any item value is set/changed and applied we must always scan recursively for children and if any are found also apply the value to them to have it work like you want together with my requirement. Or do you have any better idea?

Reply:
> do you have any better idea?

My idea is to leave this restriction as is. I'd only change the visualisation: I'd show the inherited value instead of `0`. Indeed, it may confuse a user when he/she sets 0 and apply restores the inherited value. But it is better than to display a value set by the user (zero) but not to use it.

Reply:
@oleg68 Sorry, I'm a bit confused with your answer. Can you please clarify? Can we use the existing implementation in this PR? I'd accept that version as it's clearly expressed what happens and it's never confusing with the values.

If you instead want to also display parent value for children, it's impossible to do with only the get effective value function without the confusing behaviour of a child with 0 value. That is something I won't accept. In such case the value must instead also be applied recursively to all children at the same time when the value of a parent is applied. This basically makes the get effective value funcion questionable as the current node value always will be used. The GOOrganSettingsDialog::OnEventApply part for ToneBalance will be slightly more complex but it's not so much worse than the logic for the ReleaseTail.

Reply:
I suggest only to replace `GetToneBalanceValue()` with `GetEffectiveToneBalanceValue()` and PR as is.

The confusing behavior of 0 value still exist in both variants: `GetEffectiveToneBalanceValue()` and `GetToneBalanceValue()`, so this replacing does not make worse.

Reply:
> so this replacing does not make worse.

Yes, it does, in my opinion. But it can be fixed.

Reply:
@larspalo I disagree with the `ApplyToneBalanceRecursively()` because it makes the inverse overriding: if a user sets a value on a pipe level and then sets a value on a rank level, the rank-level toneBalance will override the pipe-level toneBalance

Reply:
Exactly the same would happen with your previous suggestion.

Reply:
But a check could of course be made to see if a value already was set to something else than 0 and if so not write the parent value to it!

Reply:
After you have added a check for zero, the behavior became unstable: a user can override the inherited value with zero, as you intended, but if the user changes the parent value, the zero-setting at the pipe level stops working.

I think "never works" is better than "worked and stops working". Do you agree?

Reply:
>I think "never works" is better than "worked and suddenly stops working". Do you agree?

No not really. Because if you go back to the pipe level (or any child) and re-set to zero, which will work, it will be enforced and no filtering taking place. Also, if a pipe level already was set the rank value would not change it. So, basically this is as good a compromise as I can think of for your suggestions.

My first version would never write any value from parent at all but use the parent value if no child was set. It lacked the possibility to re-set the pipe to no filtering if parent used a non zero value, but it was always predictable. Your suggestion only made it even worse because the value from the parent would be written and used at child zero value as well as parent value.

Basically, I think that if this is not good enough for you, then we should completely remove the tone balance voicing from any level but the pipe level to never have any issues of this kind.

Reply:
Actually, the technically correct solution for having the tone balance at multiple levels would be to have individual filters for each "level" and cascade them... But this would both increase complexity and put more stress on the polyphony. But it would be the correct way to do it!